include/boost/geometry/core/check_for_same_type.hpp

```diff
@@ -0,0 +1,44 @@
+// Boost.Geometry (aka GGL, Generic Geometry Library)
+
+// Copyright (c) 2020 Barend Gehrels, Amsterdam, the Netherlands.
+
+// Use, modification and distribution is subject to the Boost Software License,
+// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+
+
+#ifndef BOOST_GEOMETRY_CORE_CHECK_FOR_SAME_TYPES_HPP
+#define BOOST_GEOMETRY_CORE_CHECK_FOR_SAME_TYPES_HPP
+
+
+#include <type_traits>
+
+#include <boost/geometry/core/coordinate_type.hpp>
+
+
+namespace boost { namespace geometry
+{
+
+#ifndef DOXYGEN_NO_DETAIL
+namespace detail
+{
+
+struct check_for_same_type
+{
+    template <typename Geometry1, typename Geometry2>
+    static inline void apply(Geometry1 const& , Geometry2 const& )
+    {
+      static_assert(std::is_same
+          <
+              typename geometry::coordinate_type<Geometry1>::type,
+              typename geometry::coordinate_type<Geometry2>::type
+          >::value, "Coordinate types in geometries should be the same");
+    }
+};
```

Comment:
3 problems with that:
- this tool static asserts for the same `coordinate_type` not the same type
- we already have similar tools for checking assertions about dimensions placed in a file where `geometry::dimension` is defined
- these tools are free functions

So I think this should be moved to `core/coordinate_type.hpp`, be called `assert_same_coordinate_type` or `assert_coordinate_type_equal` etc. and be implemented as a free function just like dimension utils.

Reply:
Sounds good, thank you. I'll update the PR.

Reply:
It's updated